src/devices/machine/mb8421.h

```diff
@@ -94,27 +94,34 @@ class mb8421_master_device : public device_t
 	// internal helpers
 	template<read_or_write row, bool is_right> void update_intr(offs_t offset);
 
+  u64 m_data_mask; // for 9/18/36bit rams
+  size_t m_ram_size;
+  size_t m_ram_mask;
+  size_t m_int_addr_left;
+  size_t m_int_addr_right;
+
 private:
 	devcb_write_line m_intl_callback;
 	devcb_write_line m_intr_callback;
 };
 
-// ======================> mb8421_device
+// ======================> mb8421_master_8bit_device
 
-class mb8421_device : public mb8421_master_device
+class mb8421_master_8bit_device : public mb8421_master_device
 {
 public:
-	mb8421_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock = 0);
-
-	u8 peek(offs_t offset) const { return m_ram[offset & 0x7ff]; }
+	u8 peek(offs_t offset) const { return m_ram[offset] & m_data_mask; }
 
 	void left_w(offs_t offset, u8 data);
 	u8 left_r(offs_t offset);
 	void right_w(offs_t offset, u8 data);
 	u8 right_r(offs_t offset);
 
+  void ram_w(offs_t offset, u8 data) { m_ram[offset] = data & m_data_mask; }
+  u8 ram_r(offs_t offset) { return m_ram[offset] & m_data_mask; }
```

Comment:
If you're always applying the mask when writing, applying it when reading as well is redundant and inefficient.  Also, these internal helpers shouldn't be public.